File: python/src/pmsec/util/paths.py
from __future__ import annotations

import os
import sys
from pathlib import Path


def npmrc_path(env: dict[str, str], home: Path) -> Path:
    if "NPM_CONFIG_USERCONFIG" in env:
        return Path(env["NPM_CONFIG_USERCONFIG"])
    return home / ".npmrc"


def pnpm_rc_path(env: dict[str, str], home: Path, platform: str) -> Path:
    """pnpm reads its global rc separately from ~/.npmrc; writing pnpm-only
    keys here keeps npm from warning (and, in npm 12, erroring) about unknown
    user config. pnpm respects XDG_CONFIG_HOME on every OS."""
    if "PMSEC_PNPM_CONFIG_FILE" in env:
        return Path(env["PMSEC_PNPM_CONFIG_FILE"])
    if "XDG_CONFIG_HOME" in env:
        return Path(env["XDG_CONFIG_HOME"]) / "pnpm" / "rc"
    if platform == "darwin":
        return home / "Library" / "Preferences" / "pnpm" / "rc"
    if platform == "win32":
        base = Path(env["LOCALAPPDATA"]) if "LOCALAPPDATA" in env else home / "AppData" / "Local"
        return base / "pnpm" / "config" / "rc"
    return home / ".config" / "pnpm" / "rc"


def bun_config_path(env: dict[str, str], home: Path) -> Path:
    if "BUN_CONFIG_FILE" in env:
        return Path(env["BUN_CONFIG_FILE"])
    return home / ".bunfig.toml"


def yarnrc_path(env: dict[str, str], home: Path) -> Path:
    if "YARN_RC_FILENAME" in env:
        return Path(env["YARN_RC_FILENAME"])
    return home / ".yarnrc.yml"


def cargo_config_path(env: dict[str, str], home: Path) -> Path:
    if "CARGO_HOME" in env:
        return Path(env["CARGO_HOME"]) / "config.toml"
    return home / ".cargo" / "config.toml"


def bundle_config_path(env: dict[str, str], home: Path) -> Path:
    """Bundler's global config. BUNDLE_USER_CONFIG points at the file directly;
    BUNDLE_USER_HOME points at the home dir (config lives at <home>/config).
    Both default to ~/.bundle, matching bundler's own resolution order."""
    if "BUNDLE_USER_CONFIG" in env:
        return Path(env["BUNDLE_USER_CONFIG"])
    if "BUNDLE_USER_HOME" in env:
        return Path(env["BUNDLE_USER_HOME"]) / "config"
    return home / ".bundle" / "config"


def mise_config_path(env: dict[str, str], home: Path, platform: str) -> Path:
    if "MISE_GLOBAL_CONFIG_FILE" in env:
        return Path(env["MISE_GLOBAL_CONFIG_FILE"])
    if platform == "win32":
        base = Path(env["LOCALAPPDATA"]) if "LOCALAPPDATA" in env else home / "AppData" / "Local"
        return base / "mise" / "config.toml"
    base = Path(env["XDG_CONFIG_HOME"]) if "XDG_CONFIG_HOME" in env else home / ".config"
    return base / "mise" / "config.toml"


def uv_config_path(env: dict[str, str], home: Path, platform: str) -> Path:
    if "UV_CONFIG_FILE" in env:
        return Path(env["UV_CONFIG_FILE"])
    if platform == "win32":
        base = Path(env["APPDATA"]) if "APPDATA" in env else home / "AppData" / "Roaming"
        return base / "uv" / "uv.toml"
    base = Path(env["XDG_CONFIG_HOME"]) if "XDG_CONFIG_HOME" in env else home / ".config"
    return base / "uv" / "uv.toml"


def current_platform() -> str:
    if os.name == "nt":
        return "win32"
    if sys.platform == "darwin":
        return "darwin"
    return "linux"

File: node/src/util/paths.mjs
import { homedir } from "node:os";
import { join } from "node:path";

export function npmrcPath(env = process.env, home = homedir()) {
  return env.NPM_CONFIG_USERCONFIG || join(home, ".npmrc");
}

// pnpm reads its global rc from a dedicated location separate from ~/.npmrc.
// Writing pnpm-only keys here keeps npm from warning (and, in npm 12, erroring)
// about unknown user config. pnpm itself respects XDG_CONFIG_HOME on every OS.
export function pnpmRcPath(env = process.env, home = homedir(), platform = process.platform) {
  if (env.PMSEC_PNPM_CONFIG_FILE) return env.PMSEC_PNPM_CONFIG_FILE;
  if (env.XDG_CONFIG_HOME) return join(env.XDG_CONFIG_HOME, "pnpm", "rc");
  if (platform === "darwin") return join(home, "Library", "Preferences", "pnpm", "rc");
  if (platform === "win32") {
    const base = env.LOCALAPPDATA || join(home, "AppData", "Local");
    return join(base, "pnpm", "config", "rc");
  }
  return join(home, ".config", "pnpm", "rc");
}

export function uvConfigPath(env = process.env, home = homedir(), platform = process.platform) {
  if (env.UV_CONFIG_FILE) return env.UV_CONFIG_FILE;
  if (platform === "win32") {
    const base = env.APPDATA || join(home, "AppData", "Roaming");
    return join(base, "uv", "uv.toml");
  }
  const base = env.XDG_CONFIG_HOME || join(home, ".config");
  return join(base, "uv", "uv.toml");
}

export function bunConfigPath(env = process.env, home = homedir()) {
  return env.BUN_CONFIG_FILE || join(home, ".bunfig.toml");
}

export function yarnrcPath(env = process.env, home = homedir()) {
  return env.YARN_RC_FILENAME || join(home, ".yarnrc.yml");
}

export function cargoConfigPath(env = process.env, home = homedir()) {
  if (env.CARGO_HOME) return join(env.CARGO_HOME, "config.toml");
  return join(home, ".cargo", "config.toml");
}

// Bundler's global config. BUNDLE_USER_CONFIG points at the file directly;
// BUNDLE_USER_HOME points at the home dir (config lives at <home>/config).
// Both default to ~/.bundle, matching bundler's own resolution order.
export function bundleConfigPath(env = process.env, home = homedir()) {
  if (env.BUNDLE_USER_CONFIG) return env.BUNDLE_USER_CONFIG;
  if (env.BUNDLE_USER_HOME) return join(env.BUNDLE_USER_HOME, "config");
  return join(home, ".bundle", "config");
}

export function miseConfigPath(env = process.env, home = homedir(), platform = process.platform) {
  if (env.MISE_GLOBAL_CONFIG_FILE) return env.MISE_GLOBAL_CONFIG_FILE;
  if (platform === "win32") {
    const base = env.LOCALAPPDATA || join(home, "AppData", "Local");
    return join(base, "mise", "config.toml");
  }
  const base = env.XDG_CONFIG_HOME || join(home, ".config");
  return join(base, "mise", "config.toml");
}
